### ISSUE TITLE
HTTP/2 support

### ISSUE THREAD

**Issue by hassy** [request/request#2033]
Hello Request maintainers and community. I am creating this issue to discuss whether there's interest in implementing HTTP/2 support in Request.

I have a working fork of Request with hacked in HTTP/2 support, but it only does what I need for my use-case (the use case is load-testing - I only ever connect to and send requests to one host, and I'm not concerned with potentially maintaining multiple connections to different hosts).

Some considerations:
- Should `http2` be an explicit option? Do we want to upgrade to HTTP/2 automatically if the server supports it?
- HTTP/2 has `push` events - these may need to be exposed to the user.
- With HTTP/2 there's a persistent connection to the server which probably can't be `unref()`d safely as is the case with HTTP. The user will need to be able to close the connection manually.
- Should HTTP/2 without TLS be allowed? (TLS is not mandatory, but strongly encouraged)
- Should this be done in Request Next? (https://github.com/request/request/issues/1982)

I'd be happy to take ownership of this and write and submit a proper PR based on the discussion here.

Thanks

**Comment by simov** [comment]
Request Next is the way to go, but I'm not sure if HTTP2 should be part of @request/core.

Request core consumes the underlying Node Core HTTP client, so in this line of thinking a better starting point would be to create a new core-like module that consumes HTTP2 compatible implementation.

Most of the code in @request/core can be safely moved around, for example you can copy/paste huge chunks out of it and probably they'll still work in another context only with minor modifications. Now I'm not saying that you should copy/paste code, but what I'm trying to say is that most of the code there is loosely coupled logic that generates small bits of state for the most part. And of course there is some streaming logic there, but then again that's inherited from Node Core.

In other words even the @request/core module is only a layer on top of the underlying HTTP/Stream mechanics and for the most part it only orchestrates various parts of the HTTP spec.

In the end if the HTTP2 support can be enabled there through an option, and if the needed logic can be isolated then why not. I'm just thinking that starting from something more generic first will be easier and less error prone.

**Comment by hassy** [comment]
Great, thanks - I'll move my changes over to Request Next first then so that we have something concrete to look at :+1:

**Comment by jsha** [comment]
With HTTP/2 support landing in Node core, I'd love to see `request` supporting it too!

**Comment by mikeal** [comment]
This will need to be fixed before we can add reliable and transparent support for h2.

https://github.com/nodejs/node/issues/16617

**Comment by gn02394847** [comment]
he has been done 
nodejs/node#16617

**Comment by paambaati** [comment]
@hassy @mikeal Can we get this started again? HTTP/2 support is in core, and `ref()` and `unref()` on client sessions is also implemented.

**Comment by OrangeDog** [comment]
I would like the ability to forward an http/2 request to http/1.1.
This current solution generates the error:
> TypeError: Header name must be a valid HTTP Token [":method"]

```js
http2.createSecureServer({allowHTTP1: true, ...}, (req, res) => {
    res.pipe(request(proxyTarget)).pipe(res);
});
```

**Comment by motin** [comment]
With Node v10 the http2 module has become a stable addition. Any known blockers for implementing HTTP/2 now?

**Comment by NathanRSmith** [comment]
Where is this "Request Next" that is referred to? Or has the situation changed since then?

**Comment by BorntraegerMarc** [comment]
@NathanRSmith did you figure out what "Request Next" is?

**Comment by NathanRSmith** [comment]
I think it was referring to https://github.com/request/request/issues/1982, but I didn't find a definitive and moved on to other things. Hoping HTTP/2 support gets added at some point.

**Comment by icodeforlove** [comment]
Would be great to see this happen.

**Comment by snowdream** [comment]
HTTP/2 support  is very useful, Any update?
